File: questao_05/questao_05.java
import java.util.ArrayList;
import java.util.List;

abstract class Animal {
    public abstract void som();
}

class Cachorro extends Animal {
    @Override
    public void som() {
        System.out.println("O cachorro faz: Au Au!");
    }
}

class Gato extends Animal {
    @Override
    public void som() {
        System.out.println("O gato faz: Meow!");
    }
}

public class questao_05 {
    public static void fazerSom(List<Animal> animais) {
        for (Animal animal : animais) {
            animal.som();
        }
    }

    public static void main(String[] args) {
        List<Animal> animais = new ArrayList<>();
        animais.add(new Cachorro());
        animais.add(new Gato());

        fazerSom(animais);
    }
}

File: questao_05/questao_05.py
# Polimorfismo: Utilize polimorfismo para criar uma função que receba uma lista de objetos "Animal" e chame o método som de cada um, demonstrando comportamento diferente para cada subclasse.

class Animal:
    def __init__(self, nome):
        self.nome = nome

    def som(self):
        raise NotImplementedError("Este método deve ser implementado nas subclasses")

class Cachorro(Animal):
    def som(self):
        return f"{self.nome} faz: Au Au!"

class Gato(Animal):
    def som(self):
        return f"{self.nome} faz: Meow!"

class Passaro(Animal):
    def som(self):
        return f"{self.nome} faz: Piu Piu!"

def emitir_sons(animais):
    for animal in animais:
        print(animal.som())

animais = [
    Cachorro("Angel"),
    Gato("Sushi"),
    Passaro("Ximi")
]

emitir_sons(animais)
